[AIRFLOW-922] Update PrestoHook to enable synchronous execution

patrickmckenna opened the pull request:
### JIRA
https://issues.apache.org/jira/browse/AIRFLOW-922

### Description
This updates [`PrestoHook`](https://github.com/apache/incubator-airflow/blob/b55f41f2c22e210d130a0b42586f0385bd5515a7/airflow/hooks/presto_hook.py) so that it can block until a statement finishes executing. Currently, [`PrestoHook.run`](https://github.com/apache/incubator-airflow/blob/b55f41f2c22e210d130a0b42586f0385bd5515a7/airflow/hooks/presto_hook.py#L104-L108) returns as soon as it sends a statement, so Operators that use it won't [run synchronously](https://github.com/apache/incubator-airflow/blob/b55f41f2c22e210d130a0b42586f0385bd5515a7/airflow/models.py#L1788-L1789).

There are other, minor changes too that seemed worth making (hence the separate commits), though they're unrelated to the primary goal of this PR. Of course if you'd rather jettison those, or put them in a separate PR, that's fine by me.

### Tests
I added [some](https://github.com/apache/incubator-airflow/blob/58ed2995867d9dc89b589270642849b6841b4b9b/tests/contrib/hooks/presto_hook_test.py) under `tests/contrib/hooks`, though I wasn't sure if that was the proper location. (`PrestoHook` is built in, not user-contributed, but AFAICT there are no existing tests for it.) 

### Commits
I haven't squashed commits yet, because I wanted to keep the history easy to see until the code is in a satisfactory state. Once it is, I'm happy to rewrite the commit history (unless you plan to just squash and merge to take care of that?).

patrickmckenna committed:
enable PrestoHook.run to execute statements synchronously

This will allow PrestoHook to be used by Operators derived from BaseOperator,
which (https://git.io/fhamQ)

  should perform or trigger certain tasks synchronously (wait for completion)

Notes on the other differences between this and DbApiHook.run:
  - no need for utf-8 encoding (https://git.io/vD9LI) b/c PyHive does it
    automatically (https://git.io/vD9Lm)
  - no closing/commiting the cursor/conn (https://git.io/vD9Lc), because those
    are no-ops w/ PyHive (https://git.io/vD9L6, https://git.io/vD9L1)

presto.Cursor does have a _poll_interval attribute, but it has no public
accessor, so it seemed safer to make that value a parameter to pass to PrestoHook.run.

patrickmckenna committed:
add minimal unit tests

patrickmckenna committed:
remove unused imports

patrickmckenna committed:
handle errors while checking execution status

patrickmckenna committed:
make names, setup a bit more explicit

mik-laj commented on airflow/hooks/presto_hook.py:
```diff
@@ -117,11 +118,49 @@ def get_pandas_df(self, hql, parameters=None):
+            return cursor.poll() is None
+        except Exception as ex:
+            msg = "Couldn't determine statement execution status: ".format(ex)
+            self.log.error(msg)
```
You should pass unformatted text to logger. 
```suggestion
            self.log.error("Couldn't determine statement execution status: %s", ex)
```

patrickmckenna replied:
@mik-laj ah, is that the agreed upon style preference? Happy to change it, just wasn't aware (didn't see anything in the docs or linting tests enforcing that, but may very well have missed it 😄).

mik-laj replied:
This is not the style of the code, but the principle of using loggers. If you format before pass to the logger, you create an unnecessary object that the logger can ignore when the level is too low. One of the ways to increase application performance is to reduce the number of logs collected. 

In special cases, the logger does not format the text. It can save the message separately and separate data to make analysis easier. If you format data before pass it to the logger, this functionality disappears. 

These notes applies to any programming language.

patrickmckenna committed:
more better exception handling, str processing

Catch only network-related exceptions when polling Presto. And make str
handling work in Python 2, too.

patrickmckenna committed:
consolidate mock setup in setUp

patrickmckenna committed:
avoid mocking time.sleep directly

patrickmckenna commented on airflow/hooks/presto_hook.py:
```diff
@@ -117,11 +120,63 @@ def get_pandas_df(self, hql, parameters=None):
+            cursor.execute(stmt, parameters)
+
+            if poll_interval is not None:
+                while not self.execution_finished(cursor):
```
It's definitely A Bad Idea™️  to poll indefinitely, but I'm assuming all the timeout logic is expected to live in the operators using this hook. @Fokko please LMK if that's assumption's inaccurate, and if there ought to be some minimal safeguards here, e.g. an upper bound on the number of pings to send or time to wait.

Fokko replied:
In other operators there are similar constructions that define an upper bound to thrown an exception when the maximum time is exceeded: https://github.com/apache/airflow/blob/master/airflow/hooks/druid_hook.py#L90
